### PR TITLE
Extract stringy Name parsing/printing printing functions to Unison.Syntax.Name module

### PR DESCRIPTION
**Pull request by mitchellwrosen**
Depends on:
- [x] #3521 

---

## Overview

This PR continues the work of decoupling the current stringy syntax of names from the data structures themselves, for the primary purpose of simplifying the implementation of fixing #1589. (Also, it is a small step in the direction of supporting alternative syntaxes).

In this PR:

- Functions that convert between `Name` and `Text` (e.g. `fromString`, `toText`, etc) have been removed from `Unison.Name` into a new module, `Unison.Syntax.Name`, which imports `Unison.Name`, and ascribes to it our one stringy syntax, which puts `.` characters between name segments, allows certain characters in name segments (whether symboly or wordy), etc).
- Ditto for `HashQualified` and `HashQualified'`, which are often fully realized as `HashQualified Name`, but sometimes `HashQualified NameSegment`.
- The `Show` instance of `Name`, `HashQualified`, and `HashQualified'` was switched to a derived `Show` instance. This is appropriate for debug messages and the like, whereas the previous instance `show = toText` really wasn't. All call sites of `show` at all types whose `Show` instances were changed were audited; debug messages were mostly left using `show`, and other user-facing things were changed to call e.g. `Name.toString` explicitly instead (with `import qualified Unison.Syntax.Name as Name`).

There is still more to do along these lines, but the PR is getting large, so I think it makes sense to cut it here, get CI passing, and review.

### CODE REVIEW

File: lib/unison-pretty-printer/src/Unison/Util/Pretty.hs [hunk 1]
```diff
@@ -722,7 +722,7 @@ wrapColumn2 rows = lines (align rows)
 
 -- Pad with enough space on the right to make all rows the same width
 leftJustify ::
-  (Eq s, Show s, LL.ListLike s Char, IsString s) =>
```
Comment (mitchellwrosen) [inline lib/unison-pretty-printer/src/Unison/Util/Pretty.hs:729]: This constraint was unused.

File: unison-src/transcripts-using-base/utf8.output.md [hunk 2]
```diff
@@ -5,9 +5,9 @@ Unison has function for converting between `Text` and a UTF-8 `Bytes` encoding o
 ```ucm
 .> find Utf8
 
-  1. builtin.Text.toUtf8 : Text -> Bytes
-  2. Text.fromUtf8 : Bytes ->{Exception} Text
-  3. builtin.Text.fromUtf8.impl : Bytes -> Either Failure Text
+  1. builtin.Text.fromUtf8.impl : Bytes -> Either Failure Text
+  2. builtin.Text.toUtf8 : Text -> Bytes
+  3. Text.fromUtf8 : Bytes ->{Exception} Text
```
Comment (mitchellwrosen) [inline unison-src/transcripts-using-base/utf8.output.md:11]: I think I might've fixed a bug here accidentally. Isn't the new order better?

File: parser-typechecker/src/Unison/Result.hs [hunk 3]
```diff
@@ -30,7 +27,6 @@ data Note v loc
   | TypeError (Context.ErrorNote v loc)
   | TypeInfo (Context.InfoNote v loc)
   | CompilerBug (CompilerBug v loc)
-  deriving (Show)
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Result.hs]: Hrmm, does this have to be removed? Show instances are handy for debugging.

File: parser-typechecker/src/Unison/UnisonFile/Names.hs [hunk 4]
```diff
@@ -31,8 +26,8 @@ import qualified Unison.WatchKind as WK
 toNames :: Var v => UnisonFile v a -> Names
 toNames uf = datas <> effects
   where
-    datas = foldMap DD.Names.dataDeclToNames' (Map.toList (UF.dataDeclarationsId uf))
-    effects = foldMap DD.Names.effectDeclToNames' (Map.toList (UF.effectDeclarationsId uf))
+    datas = foldMap (DD.Names.dataDeclToNames' Name.unsafeFromVar) (Map.toList (UF.dataDeclarationsId uf))
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/UnisonFile/Names.hs:29]: Just curious, what would be involved in adding a non-empty guarantee to Vars so that this isn't unsafe?

File: unison-cli/src/Unison/Codebase/Editor/Output.hs [hunk 5]
```diff
@@ -179,8 +179,8 @@ data Output
   | ListNames
       IsGlobal
       Int -- hq length to print References
-      [(Reference, Set (HQ'.HashQualified Name))] -- type match, type names
-      [(Referent, Set (HQ'.HashQualified Name))] -- term match, term names
+      [(Reference, [HQ'.HashQualified Name])] -- type match, type names
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/Output.hs:182]: Just curious, why this change?

File: unison-cli/src/Unison/Codebase/Editor/Output/BranchDiff.hs [hunk 6]
```diff
@@ -88,8 +86,16 @@ isEmpty BranchDiffOutput {..} =
 
 type TermDisplay v a = (HashQualified Name, Referent, Maybe (Type v a), MetadataDiff (MetadataDisplay v a))
 
+compareTermDisplay :: TermDisplay v a -> TermDisplay v a -> Ordering
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/Output/BranchDiff.hs:89]: Why not use an Ord instance here btw?
If this isn't meant to be the canonical ord instance, then maybe add some context to the name about what it's comparing and why

File: unison-core/src/Unison/DataDeclaration/Names.hs [hunk 7]
```diff
@@ -24,35 +18,36 @@ import Unison.Var (Var)
 import Prelude hiding (cycle)
 
 -- implementation of dataDeclToNames and effectDeclToNames
-toNames :: Var v => CT.ConstructorType -> v -> Reference.Id -> DataDeclaration v a -> Names
-toNames ct typeSymbol (Reference.DerivedId -> r) dd =
+toNames :: Var v => (v -> Name.Name) -> CT.ConstructorType -> v -> Reference.Id -> DataDeclaration v a -> Names
+toNames unsafeVarToName ct typeSymbol (Reference.DerivedId -> r) dd =
```
Comment (ChrisPenner) [inline unison-core/src/Unison/DataDeclaration/Names.hs]: Whether it's unsafe is moreso up to the caller right?

```suggestion
toNames varToName ct typeSymbol (Reference.DerivedId -> r) dd =
```

File: unison-core/src/Unison/NamesWithHistory.hs [hunk 8]
```diff
@@ -71,7 +71,6 @@ data Diff = Diff
     addedNames :: Names,
     removedNames :: Names
   }
-  deriving (Show)
```
Comment (ChrisPenner) [inline unison-core/src/Unison/NamesWithHistory.hs]: Just curious why removing this? I see you're still deriving `Show` for `Names` above.

**Review by ChrisPenner — APPROVED** [review]
Noted that I'd like to keep as many `Show` instances as we can for debugging purposes, but otherwise looks good

File: unison-cli/src/Unison/Codebase/Editor/Output/BranchDiff.hs (hunk 6 above)
Comment (mitchellwrosen) [reply unison-cli/src/Unison/Codebase/Editor/Output/BranchDiff.hs:89]: The derived `Ord` instance previously relied on being able to render a `Name` to a string, so it won't do. These comparisons just recover the existing behavior of comparing names alphabetically.

Although - now that you've pointed it out - I think we could just rework the `instance Alphabetical Name` to not rely on a `toText`. It's currently:

```
instance Alphabetical Name where
  compareAlphabetical n1 n2 =
    compareAlphabetical (toText n1) (toText n2)
```

but it'd probably be better to just compare the _segments_' bytes alphabetically. Then I think we could just use the derived `Ord` again.

File: unison-cli/src/Unison/Codebase/Editor/Output.hs (hunk 5 above)
Comment (mitchellwrosen) [reply unison-cli/src/Unison/Codebase/Editor/Output.hs:182]: The `Set` was only made at the last instant to get the names in the order we want to display them. But Name's `Ord` instance no longer does that, it's just derived.

File: parser-typechecker/src/Unison/UnisonFile/Names.hs (hunk 4 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/UnisonFile/Names.hs:29]: I think the `Var` <-> `Name` conversions need a little (or a lot) more elbow grease before we can have a safe conversion. Right now I think it's only really sensible to convert a `Var` into a `Name` if it happens to have `User` type, which was actually typed in by a user (and so looks something like `.foo.bar.baz`). But we have a couple places where we stuff non-user-written things into the `User` type (like `User "()"`, for foralls).

File: parser-typechecker/src/Unison/Result.hs (hunk 3 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Result.hs]: No; adding it (and others) back. I had initially removed them all only to audit all call sites to all of the `Show` instances that were transitively depending on `Name`'s show, to switch them over to calling `toText` instead. They could/should be added back (although I will note that `recover-rtti` seems a lot better for debugging to me - no need to muscle through `Show` instances everywhere, including on types that can't have a `Show` instance for whatever reason)

**Comment by mitchellwrosen** [comment]
@ChrisPenner I think I'll merge this for now and follow up with moving the `instance Alphabetical Name`, so we can delete those custom `compareTermDisplay` things.
